File: Data Structures & Algorithms/three-integer-sum/submission-2.kt
class Solution {
    fun threeSum(nums: IntArray): List<List<Int>> {
        nums.sort()
        val result = mutableListOf<List<Int>>()

        for ((index, num) in nums.withIndex()) {
            if (index > 0 && nums[index - 1] == num) continue

            var left = index + 1
            var right = nums.size - 1
            while (left < right) {
                val threeSum = num + nums[left] + nums[right]
                when {
                    threeSum > 0 -> right--
                    threeSum < 0 -> left++
                    threeSum == 0 -> {
                        result.add(listOf(num, nums[left], nums[right]))
                        left++
                        right--
                        while (left < right && nums[left] == nums[left - 1]) {
                            left++
                        }
                        while (left < right && nums[right] == nums[right + 1]) {
                            right--
                        }
                    }
                }
            }
        }

        return result
    }
}

File: Data Structures & Algorithms/longest-palindromic-substring/submission-2.py
class Solution:
    def longestPalindrome(self, s: str) -> str:
        def is_palindrome(s: str):
            return s == s[::-1]
        
        result = ""
        for index, letter in enumerate(s):
            left = right = index
            while left >= 0 and right <= len(s) - 1 and is_palindrome(s[left:right + 1]):
                inner = s[left:right + 1]
                result = inner if len(inner) > len(result) else result
                left -= 1
                right += 1

            left = index
            right = index + 1

            while left >= 0 and right <= len(s) - 1 and is_palindrome(s[left:right + 1]):
                inner = s[left:right + 1]
                result = inner if len(inner) > len(result) else result
                left -= 1
                right += 1

        return result